mesh/mesh.go

```diff
@@ -176,42 +181,55 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
 	m.lvMutex.Unlock()
 
 	for layerId := oldPbase; layerId < newPbase; layerId++ {
-		m.AccumulateRewards(layerId, m.config)
-		if err := m.PushTransactions(layerId); err != nil {
-			m.With().Error("failed to push transactions", log.Err(err))
+		l, err := m.GetLayer(layerId)
+		if err != nil || l == nil {
+			// TODO: propagate/handle error
+			m.With().Error("failed to get layer", log.LayerId(layerId.Uint64()), log.Err(err))
 			break
 		}
+		m.AccumulateRewards(l, m.config)
+		m.PushTransactions(l)
+		m.logStateRoot(layerId)
+		m.setLayerHash(l)
 	}
+	m.persistLayerHash()
 	m.Info("done validating layer %v", lyr.Index())
 }
 
+func (m *Mesh) logStateRoot(layerId types.LayerID) {
+	m.Event().Info("end of layer state root",
+		log.LayerId(layerId.Uint64()),
+		log.String("state_root", util.Bytes2Hex(m.TxProcessor.GetStateRoot().Bytes())),
+	)
+}
+
+func (m *Mesh) setLayerHash(layer *types.Layer) {
+	validBlocks, _ := m.BlocksByValidity(layer.Blocks())
+	m.layerHash = types.CalcBlocksHash32(types.BlockIds(validBlocks), m.layerHash).Bytes()
+
+	m.Event().Info("new layer hash",
+		log.LayerId(layer.Index().Uint64()),
+		log.String("layer_hash", util.Bytes2Hex(m.layerHash)))
+}
+
+func (m *Mesh) persistLayerHash() {
+	if err := m.general.Put(LAYERHASH, m.layerHash); err != nil {
+		m.With().Error("failed to persist layer hash", log.Err(err), log.LayerId(m.validatedLayer.Uint64()),
+			log.String("layer_hash", util.Bytes2Hex(m.layerHash)))
+	}
+}
+
 func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) (validBlockTxs, invalidBlockTxs []*types.Transaction) {
 	// Separate blocks by validity
-	var validBlocks, invalidBlocks []*types.Block
-	for _, b := range l.Blocks() {
-		valid, err := m.ContextualValidity(b.Id())
-		if err != nil {
-			m.With().Error("could not get contextual validity", log.BlockId(b.Id().String()), log.Err(err))
-		}
-		if valid {
-			validBlocks = append(validBlocks, b)
-		} else {
-			invalidBlocks = append(invalidBlocks, b)
-		}
-	}
+	validBlocks, invalidBlocks := m.BlocksByValidity(l.Blocks())
 
 	// Deterministically sort valid blocks
-	validBlocks = types.SortBlocks(validBlocks)
-
-	// `layerHash` is the sha256 sum of sorted layer block IDs
-	layerHash := sha256.New()
-	for _, b := range validBlocks {
-		layerHash.Write(b.Id().AsHash32().Bytes())
-	}
+	types.SortBlocks(validBlocks)
```

Comment:
why not use CalcBlocksHash32 ?

Reply:
I need to have to blocks sorted for later in the method

Reply:
yes but CalcBlocksHash32 also sorts the blocks (it has to)

Reply:
cool i get that your trying to save on the sort, but i would consider sorting twice and making CalcBlockHash32Presorted private just to have leaner code and maybe prevent someone using it wrong in the future.
sorting the 200 ids twice is negligible in my opinion

Reply:
also CalcBlockHash32Presorted sounded to me like the method expects a presorted slice, might just be me though ...

Reply:
It expects a pre sorted slice. What makes you think otherwise?

Reply:
Maybe you’re right. It felt more “correct” to me this way. 

I’ll take another look at it tomorrow.

Reply:
I'm not sure about this.

This is copying and sorting 6.4 KB using a rather expensive comparison function. Not a big deal, but if we can easily avoid it, why not?

I'm actually considering doing the opposite... It looks like all other usages of `CalcBlocksHash32()` don't need the copy of the slice, they pass a slice that's only used for this method. So we could remove `CalcBlocksHash32()` and then just call something like `CalcBlockHash32Presorted(SortBlockIds(ids))`.

What do you think?

Reply:
i usually prefer code that is less prone to bad use, also sorting an already sorted array is very fast, and the compare compare bytes slices doesn't seem expensive at all

Reply:
`CalcBlockHash32Presorted()` is prone to bad use? I disagree on that.